File: strategic-jl/src/core/provenance.jl
struct ProvenanceNode
    id::Union{String, Nothing}              # Optional UUID v4
    operation::String
    trait_type::Union{String, Nothing}      # Populated iff operation == "applied_trait"
    chapter_ref::String
    theoretical_origin::Union{String, Nothing}
    rationale::String
    parent_id::String
    timestamp::DateTime
    author::Symbol                          # :user, :llm, :system
end

"""
    ProvenanceNode(operation, chapter_ref, rationale; kwargs...)

Construct a provenance node. Generates a UUID v4 by default so downstream
layers (LLM explanations, web composer) can reference the node stably;
pass `id=nothing` to opt out.
"""
function ProvenanceNode(op::String, chapter::String, rationale::String;
        id::Union{String, Nothing} = string(uuid4()),
        trait_type::Union{String, Nothing} = nothing,
        parent_id::String = "",
        theoretical_origin::Union{String, Nothing} = nothing,
        author::Symbol = :user)
    ProvenanceNode(id, op, trait_type, chapter, theoretical_origin,
        rationale, parent_id, now(), author)
end

function append_provenance!(world::StrategicWorld, node::ProvenanceNode)
    push!(world.provenance, node)
    world
end

File: strategic-rs/crates/strategic-core/src/provenance.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProvenanceNode {
    /// Optional UUID v4. Populated when the producing engine wants
    /// downstream layers (LLM explanations, web composer) to be able
    /// to reference this specific node by id.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    pub operation: String,
    /// Populated iff `operation == "applied_trait"`. One of the
    /// TraitType enum values. Split from `operation` so explanation
    /// layers can group by trait family without string parsing.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trait_type: Option<String>,
    pub chapter_ref: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub theoretical_origin: Option<String>,
    #[serde(default)]
    pub rationale: String,
    pub parent_id: String,
    pub timestamp: String,
    pub author: Author,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Author {
    User,
    Llm,
    System,
}
